### PR TITLE
Add 'isNamedDeclaration' helper to reduce casts

### PR DESCRIPTION
**Pull request by andy-ms**
Especially helps with #22088 because testing `(node as NamedDeclaration).name` doesn't make `(node as NamedDeclaration).name` defined in future uses.

### CODE REVIEW

File: src/compiler/utilities.ts [hunk 1]
```diff
@@ -4298,6 +4298,11 @@ namespace ts {
         return declaration.name || nameForNamelessJSDocTypedef(declaration);
     }
 
+    /** @internal */
+    export function isNamedDeclaration(node: Node): node is NamedDeclaration & { name: DeclarationName } {
```
Comment (DanielRosenwasser) [inline src/compiler/utilities.ts]: Mmm, this implementation doesn't seem to actually check that the name is a `DeclarationName`.

Comment (andy-ms) [reply src/compiler/utilities.ts]: Added an assertion -- a `name` property should always be a `DeclarationName`.

File: src/compiler/utilities.ts [hunk 2]
```diff
@@ -4298,6 +4298,13 @@ namespace ts {
         return declaration.name || nameForNamelessJSDocTypedef(declaration);
     }
 
+    /** @internal */
+    export function isNamedDeclaration(node: Node): node is NamedDeclaration & { name: DeclarationName } {
+        const name = (node as NamedDeclaration).name;
+        Debug.assert(!name || isDeclaration(node) && node.name === name, "A 'name' property should always be a DeclarationName.");
```
Comment (mhegazy) [inline src/compiler/utilities.ts]: this really a hot path, i would rather not do additional work. this is rather simple check. does it have a `name` property or not.

**Comment by cowchimp** [comment]
Hi. is this new Type Guard not exposed externally as all of the other Type Guards?
I'm using Typescript 2.8.3 and it's not listed inside of `lib/typescript.d.ts` so I get a compiler error when I try to use it in my code.
Thanks.

**Comment by mhegazy** [comment]
You can send us a PR to remove the `/** @internal */` comment

**Pull request merged**